### ISSUE TITLE
Dynamic routing based on username

### ISSUE THREAD

**Issue by matpen** [oauth2-proxy/oauth2-proxy#1785]
For my use case, I would need a way to route requests dynamically based on username (and possibly other token features). For example, user A will be routed to `http://127.0.0.1:8001` and user B will be routed to `http://127.0.0.1:8002`.

## Expected Behavior

The idea is that `upstream` is a single-user, auth-unaware application (e.g. Jupyter). Multiple instances of the applications are running as separated backends, listening on different addresses. There is therefore the need for a reverse proxy system with two functionalities:
* authentication (already provided by `oauth-proxy`)
* dynamic routing (not available, as far as I can tell)

## Current Behavior

Perhaps this could be implemented by extending the [--upstream setting](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/#config-file) from a list to a map, but I do not see any references of such a functionality in the docs.

## Possible Solution

As a temporary workaround I thought about chaining `oauth-proxy` to another reverse proxy, but after exploring the Internet, it seems like the solutions for implementing dynamic routing range from very difficult to impossible: example documents for [nginx](https://stackoverflow.com/questions/45886007/nginx-dynamic-routing-per-browser-user-session), [HAproxy](https://serverfault.com/questions/771477/haproxy-dynamic-server-address-based-off-of-header-value), [curity](https://curity.io/resources/learn/implementing-dynamic-user-routing/). Istio seems to have a [workable implementation](https://istio.io/latest/docs/tasks/traffic-management/request-routing/), but its installation seems overkill for the goal. 

## Steps to Reproduce (for bugs)

n / a

## Context

See above

## Your Environment

kubernetes

**Comment by JoelSpeed** [comment]
Is it not possible to handle this in your backend? If a user tries to access a page they shouldn't have access to, redirect to the correct page?

I don't think this is something we are likely to want to add soon, we have other higher priorities at the moment

**Comment by matpen** [comment]
@JoelSpeed thank you for your reply. Of course it is possible to do this in application code, but if one does that, then they will also handle authentication, too. That is indeed what I am doing at the moment, and it works well, but it is cumbersome to maintain, because the upstream is a third-party app that cannot be modified, so the backend must be one independent deployment to handle auth + routing.

In other words, due to this missing feature, `oauth2-proxy` is not applicable for this use case, and one must resort to other tools (or writing them from scratch).

**Comment by JoelSpeed** [comment]
Yep, happy to keep this on the backlog for the future :+1:

**Comment by therc** [comment]
Since you say that you want to run this under Kubernetes, the most appropriate place to do this would be in an Ingress, which would then delegate authentication to oauth2-proxy. The nginx ingress controller does support the latter, but not the former. I did file a feature request months ago: https://github.com/kubernetes/ingress-nginx/issues/7498 
Basically, you could put all your pods behind a headless Service, as long as you also set a pod hostname/subdomain (so that the proxy has a predictable upstream address to forward to). Then you'd need a wildcard Ingress as well. User A would hit A.jupyter.mydomain.com, which would be forwarded to host (pod) A.jupyter.svc.cluster.local, etc. You can block user B from hitting A.jupyter with some custom Lua code in the ingress annotation.
The feature request was closed for inactivity, but I'm tempted to just take an existing load balancing method's code and repurpose it (I don't really know Lua, we'll see...)

**Comment by matpen** [comment]
@therc thank you for your input: however, it does not seem scalable to me, because for every user one would need a subdomain. That said, a solution based on a combination of `oauth2-proxy` and `nginx` (or any other reverse proxy) with some scripting help could be viable, and could probably cover both the "dynamic" as well as the "subdomain based" approaches. 

I did some research in this direction, as specified in the OP, but I am facing the same problem: I do not speak Lua. Also, it looks like you need a plugin for that to work under `nginx`, which means custom images, etc.

**Comment by matpen** [comment]
Interesting, thank you for the explanation, @therc 

> So it's "just" a matter of writing the code and getting it accepted upstream.
> The feature request was closed for inactivity, but I'm tempted to just take an existing load balancing method's code and repurpose it (I don't really know Lua, we'll see...)

If you give in to the "temptation", would you remember to ping in this issue? I will definitely be backing you if we need to show the upstream developers that the feature is needed.

**Comment by matpen** [comment]
@therc can we disable the stale bot on this?

**Comment by matpen** [comment]
I am still interested in this issue.
